src/compas/files/_xml/xml_pre_38.py

```diff
@@ -0,0 +1,73 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import xml.etree.ElementTree as ET
+from xml.dom import NotSupportedErr
+
+from compas.files._xml.xml_cpython import prettify_string  # doesn't need special handling for pre-3.8 so we just import
+
+__all__ = [
+    'xml_from_file',
+    'xml_from_string',
+    'prettify_string',
+]
+
+
+def xml_from_file(source, tree_parser=None):
+    if tree_parser:
+        raise NotSupportedErr('XML parsing on CPython 3.7 and older does not support a custom tree parser')
+
+    tree_parser = ET.XMLPullParser
+    parser = tree_parser(events=('start', 'end', 'start-ns', 'end-ns'))
+
+    close_source = False
+    if not hasattr(source, 'read'):
+        source = open(source, 'rb')
+        close_source = True
+    try:
+        while True:
+            data = source.read(65536)
+            if not data:
+                break
+            parser.feed(data)
+
+        return _process_all_events(parser)
+    finally:
+        if close_source:
+            source.close()
```

Comment:
I think we could get some mileage out of abstracting this open-if-necessary-and-read-resulting-file-like-object into a context manager of our own stashed away in `compas.utilities` and then this code could be smushed to `open_file(source, 'rb', close_source=True)`.

Reply:
good point, but I will split that change into its own PR